File: 29-divide-two-integers/29-divide-two-integers.py
class Solution:
    def divide(self, dividend: int, divisor: int) -> int:
        MIN_INT = -2**31
        MAX_INT = 2**31 - 1
        print(MIN_INT)
        print(MAX_INT)
        if dividend == MIN_INT and divisor == -1:
            return MAX_INT
        
        negative = (dividend < 0 and divisor > 0) or (dividend > 0 and divisor < 0)
        dividend, divisor = abs(dividend), abs(divisor)
        shift, res = 31, 0
        
        while dividend >= divisor:
            while dividend < (divisor << shift):
                shift -= 1
            dividend -= divisor << shift
            res += 1 << shift
        return res if not negative else -res

File: 29-divide-two-integers/29-divide-two-integers.java
class Solution {
    public int divide(int x, int y) {
        List<Long> bin = new ArrayList<>();
        boolean neg = false;
        if((x > 0 && y < 0) || (x < 0 && y > 0)) neg = true;
        long a = Math.abs((long)x), b = Math.abs((long)y); 
        for(long i = b; i <= (long)a; i = i+ i) bin.add(i);
        long res = 0;
        
        for(int i = bin.size() - 1; i >=0; i--) {
            if(a >= bin.get(i)) {
                a -= bin.get(i);
                res += 1L << i;
            }
        } 
        if(neg) res = -res; 
        if(res > Integer.MAX_VALUE || res < Integer.MIN_VALUE) return Integer.MAX_VALUE;
        return (int)res;
    }
}